### PR TITLE
UART framing API

### PR DESCRIPTION
**Pull request by andyp1per**
This implements a UART framing API and uses it in our CRSF protocol implementation.

Flown briefly using CRSF tracer on a 3" quad.

### CODE REVIEW

File: libraries/AP_HAL_ChibiOS/UARTDriver.h [hunk 1]
```diff
@@ -180,6 +180,7 @@ class ChibiOS::UARTDriver : public AP_HAL::UARTDriver {
 #endif
     ByteBuffer _readbuf{0};
     ByteBuffer _writebuf{0};
+    ByteBuffer _frame_readbuf{0};
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/UARTDriver.h:183]: is the intention that _frame_readbuf holds max 1 frame? Or multiple frames?
I was expecting this to be a buffer of frames, so multiple can be outstanding. We could make it just 1 though, in which case why make it a ByteBuffer?

File: libraries/AP_HAL_ChibiOS/UARTDriver.cpp [hunk 2]
```diff
@@ -478,6 +479,47 @@ void UARTDriver::_begin(uint32_t b, uint16_t rxS, uint16_t txS)
     }
 }
 
+void UARTDriver::_begin_framing(uint16_t _frame_size, uint16_t frames_in_buffer)
+{
+    _frame_readbuf.clear();
+    _frame_readbuf.set_size((_frame_size + sizeof(uint32_t)) * frames_in_buffer);
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/UARTDriver.cpp:485]: why the +4?

File: libraries/AP_HAL_ChibiOS/UARTDriver.cpp [hunk 3]
```diff
@@ -529,12 +569,46 @@ void UARTDriver::dma_tx_deallocate(Shared_DMA *ctx)
...
+    // framing error, restart at the next frame
+    if (fsize > frame_size) {
+        _frame_readbuf.clear();
+        _readbuf.clear();
+        return;
+    }
+
+    // read the remaining available bytes as a frame
+    _frame_readbuf.write((const uint8_t*)&fsize, sizeof(uint32_t));
+    while (fsize > 0) {
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/UARTDriver.cpp:593]: nasty problem if another thread does a read, we could end up with fsize wrapping, and we fall out of the sky with the while loop

**Review by tridge — CHANGES_REQUESTED** [review]
where is the metadata that tells us how long each frame is?

File: libraries/AP_HAL/UARTDriver.h [hunk 4]
```diff
@@ -200,6 +218,13 @@ class AP_HAL::UARTDriver : public AP_HAL::BetterStream {
      */
     virtual ssize_t _read(uint8_t *buffer, uint16_t count)  WARN_IF_UNUSED = 0;
 
+      /*
+      frame-based read
+     */
+    virtual ssize_t _read_frame(uint8_t *buffer, uint16_t count) {
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply libraries/AP_HAL/UARTDriver.h:233]: its the size of buffer. added some comments

File: libraries/AP_HAL/UARTDriver.h [hunk 5]
```diff
@@ -200,6 +218,13 @@ class AP_HAL::UARTDriver : public AP_HAL::BetterStream {
...
+      return _read(buffer, count < frame_size ? count : frame_size);
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply libraries/AP_HAL/UARTDriver.h:235]: its just trying to provide something useful on HALs that don't support the underlying framing mechanism

File: libraries/AP_HAL_ChibiOS/UARTDriver.cpp (hunk 2 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/UARTDriver.cpp:489]: It's the length of the frame. frames can be variable length remember - the size is just the maximum size

File: libraries/AP_HAL_ChibiOS/UARTDriver.cpp (hunk 3 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/UARTDriver.cpp:608]: I've now locked the kernel around this call - I don't think another thread can read until we are done

Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/UARTDriver.cpp:608]: I've now locked the kernel around this call so I don't think another thread can read while the transfer is in progress

File: libraries/AP_HAL/UARTDriver.cpp [hunk 6]
```diff
@@ -111,6 +132,27 @@ bool AP_HAL::UARTDriver::read(uint8_t &b)
     return n > 0;
 }
 
+/*
+    frame-based read. An entire frame will be read into buffer. count indicates the size of the
+    buffer and if this is less than the size of the frame to be read, the remaining bytes in the frame will
+    be discarded. returns the size of the frame that was read or 0 if no frames are available to
+    be read.
+*/
+ssize_t AP_HAL::UARTDriver::read_frame(uint8_t *buf, uint16_t count)
```
Comment (tridge) [inline libraries/AP_HAL/UARTDriver.cpp:141]: buf_size would be clearer than count

**Review by tridge — CHANGES_REQUESTED** [review]
I'd like to try a simpler version that only holds 1 frame, if a new frame arrives while we have a frame pending then we discard the old frame
I can't immediately think of a case where we need multi-frame, but if we do we could add that once this has settled in

File: libraries/AP_HAL/UARTDriver.cpp [hunk 7]
```diff
@@ -9,14 +9,35 @@ void AP_HAL::UARTDriver::begin(uint32_t baud, uint16_t rxSpace, uint16_t txSpace
...
+    begin in frame-based mode. the framing API works on top of the 
+    existing UART infrastructure and framing can be started at any
+    time after the UART has been started. Consumed frames can be at most frame_size
+    bytes long, frames longer than this will be considered corrupt and the
+    framing buffer will be reset. The framing buffer holds at most
+    frames_in_buffer frames, the frames are stored in a ring buffer and so
+    older frames that have not been read will be overwritten.
+    Returns true if the framing API could be started, false otherwise
+ */
+bool AP_HAL::UARTDriver::begin_framing(uint16_t _frame_size)
```
Comment (tridge) [inline libraries/AP_HAL/UARTDriver.cpp:30]: max_frame_size would be clearer

File: libraries/AP_HAL_ChibiOS/UARTDriver.cpp [hunk 8]
```diff
@@ -535,12 +608,43 @@ void UARTDriver::dma_tx_deallocate(Shared_DMA *ctx)
...
+void UARTDriver::transfer_frame()
+{
+    const uint32_t fsize = _readbuf.available();
+
+    // no data
+    if (fsize == 0) {
+        return;
+    }
+
+    _frame_bounce_idx ^= 1;
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/UARTDriver.cpp:627]: this should come after filling the buffer

File: libraries/AP_HAL_ChibiOS/UARTDriver.cpp [hunk 9]
```diff
@@ -535,12 +608,43 @@ void UARTDriver::dma_tx_deallocate(Shared_DMA *ctx)
...
+    _frame_bounce_len[_frame_bounce_idx] = _readbuf.read(_frame_bounce_buf[_frame_bounce_idx], fsize);
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/UARTDriver.cpp:628]: use tmp variable here

File: libraries/AP_HAL_ChibiOS/UARTDriver.h (hunk 1 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/UARTDriver.h]: Fixed to use A/B buffers

File: libraries/AP_HAL_ChibiOS/UARTDriver.cpp (hunk 2 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/UARTDriver.cpp]: Removed

File: libraries/AP_HAL_ChibiOS/UARTDriver.cpp (hunk 3 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/UARTDriver.cpp]: Removed

File: libraries/AP_HAL/UARTDriver.cpp (hunk 6 above)
Comment (andyp1per) [reply libraries/AP_HAL/UARTDriver.cpp]: Fixed

File: libraries/AP_HAL/UARTDriver.cpp (hunk 7 above)
Comment (andyp1per) [reply libraries/AP_HAL/UARTDriver.cpp]: Fixed

File: libraries/AP_HAL_ChibiOS/UARTDriver.cpp [hunk 10]
```diff
@@ -535,12 +604,52 @@ void UARTDriver::dma_tx_deallocate(Shared_DMA *ctx)
...
+    }
+
+    // discard the old frame if necessary
+    _frame_readbuf.clear();
+
+    // read the remaining available bytes as a frame
+    while (fsize > 0) {
+        uint32_t avail;
+        const uint8_t* buf = _readbuf.readptr(avail);
+        _frame_readbuf.write(buf, avail);
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/UARTDriver.cpp]: Fixed

File: libraries/AP_HAL_ChibiOS/UARTDriver.cpp (hunk 8 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/UARTDriver.cpp]: Fixed

File: libraries/AP_HAL_ChibiOS/UARTDriver.cpp (hunk 9 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/UARTDriver.cpp]: Fixed

**Comment by andyp1per** [comment]
Now working with A/B variant and cleaned up. Actually found some subtle issues with CRSFv3 and IDLE frames. Flown briefly on my 3"

**Comment by andyp1per** [comment]
This turned out to have too many problems and doesn't actually work reliably on TBS kit

**Pull request closed without merging**